kotekan/kotekan.cpp

```diff
@@ -356,15 +463,39 @@ int main(int argc, char** argv) {
     rest_server.start(address_parts.at(0), std::stoi(address_parts.at(1)));
 
     if (string(config_file_name) != "none") {
-        // TODO should be in a try catch block, to make failures cleaner.
         std::lock_guard<std::mutex> lock(kotekan_state_lock);
         INFO_NON_OO("Opening config file {:s}", config_file_name);
 
-        std::string exec_command =
-            fmt::format(fmt("python -c '{:s}' {:s}"), yaml_to_json, config_file_name);
+        // Create the format string, adding the yaml dump, and extra vars if needed
+        std::string exec_format_str = "python -c '{:s}' ";
+        if (dump_yaml) {
+            exec_format_str += "-d ";
+        }
+        if (jinja_variables != "") {
+            exec_format_str += "-e '{:s}' ";
+        }
+        exec_format_str += "{:s}";
+
+        std::string exec_command = "";
+        if (jinja_variables == "")
+            exec_command = fmt::format(exec_format_str, yaml_to_json, config_file_name);
+        else
+            exec_command =
+                fmt::format(exec_format_str, yaml_to_json, jinja_variables, config_file_name);
```

Comment:
There is a potential security vulnerability here.

If I run kotekan with a malicious value for `-e` along the lines of:
```
kotekan -e "' \`rm -rvf *\` '" [...]
```
then I can execute arbitrary shell commands as the UID of kotekan.  This becomes an escalation attack whenever kotekan is installed setuid. (I use a backtick subshell to ensure my malicious payload is executed before the command I'm co-opting.)

The root problem is the use of `popen`, which invokes `sh(1)` to parse the string it's given.  Given that it's kotekan itself that built the string, this is completely unnecessary.  It would be best to replace `popen` with the standard `fork/execvp` idiom which would remove the need to compose a sh-style command line because you'd be passing the arguments to `python` explicitly, bypassing the shell parser.

It *might* be enough to run the user-supplied string through a JSON validator to ensure valid JSON, but I suspect JSON is expressive enough that I can escape the sh-single quotes within valid JSON, so I think the only other safe option is to properly `sh`-escape the user-supplied string before creating the command-line.

Reply:
Good catch!   I don't think we have any situations where the privilege of `kotekan` is higher than the user in command line mode, but this should be fixed nonetheless.   If you want to write fork-and-exec replacement, that would be appreciated!  But if you don't have time I can do that as well, just let me know?

Reply:
sure, I'll give it a stab. I'll try to get something done today.

Reply:
Thanks!